File: algorithms/solutions/week32/manhdv/leaf_similar_trees.rb
# https://leetcode.com/problems/leaf-similar-trees/description/

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

def fetch_leaves(node, leaves)
  node.left.nil? && node.right.nil? && leaves << node.val

  fetch_leaves(node.left, leaves) unless node.left.nil?
  fetch_leaves(node.right, leaves) unless node.right.nil?

  leaves
end

# @param {TreeNode} root1
# @param {TreeNode} root2
# @return {Boolean}
def leaf_similar(root1, root2)
  fetch_leaves(root1, []) == fetch_leaves(root2, [])
end

File: algorithms/solutions/week32/taurandat/leaf-similar-trees.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  bool leafSimilar(TreeNode* root1, TreeNode* root2) {
    vector<int> sequence1, sequence2;
    traverse(root1, sequence1);
    traverse(root2, sequence2);

    return sequence1 == sequence2;
  }

  void traverse(TreeNode* node, vector<int>& sequence) {
    if(node == nullptr) {
      return;
    } else if(node->left == nullptr && node->right == nullptr) {
      sequence.push_back(node->val);
      return;
    } else {
      traverse(node->left, sequence);
      traverse(node->right, sequence);
      return;
    }
  }
};
